ACPs/118-warp-signature-request/README.md

```diff
@@ -0,0 +1,81 @@
+
+## Abstract
+
+Proposes handling Warp signature requests from peers at the protocol level, rather than leaving it to VMs to implement. This will standardize the signature request interface, allowing for more flexible, VM-agnostic signature aggregation implementations.
+
+## Motivation
+
+Warp message signatures consist of an aggregate BLS signature composed of the individual signatures of a subnet's validators. Individual signatures need to be retreivable by the party that wishes to construct an aggregate signature. At present, this is left to VMs to implement, as is the case with [Subnet EVM](https://github.com/ava-labs/subnet-evm/blob/v0.6.6/plugin/evm/message/signature_request.go#20) and [Coreth](https://github.com/ava-labs/coreth/blob/v0.13.6-rc.0/plugin/evm/message/signature_request.go#L20)
+
+This creates friction in applications that are intended to operate across many VMs (or distinct implementations of the same VM). As an example, the reference Warp message relayer implementation, [awm-relayer](https://github.com/ava-labs/awm-relayer), fetches individual signatures from validators and aggregates them before sending the Warp message to its destination chain for verification. However, Subnet EVM and Coreth have distinct codecs, requiring the relayer to [switch](https://github.com/ava-labs/awm-relayer/blob/v1.4.0-rc.0/relayer/application_relayer.go#L372) according to the target codebase.
+
+Another example is ACP-75, which aims to implement acceptance proofs using Warp. The signature aggregation mechanism is not [specified](https://github.com/avalanche-foundation/ACPs/blob/main/ACPs/75-acceptance-proofs/README.md#signature-aggregation), which is a blocker for that ACP to be marked implementable.
+
+Standardizing the Warp Signature Request format by implementing it as a protocol-level (AvalancheGo) AppRequest type would simplify the implementation of ACP-75, and streamline signature aggregation for out-of-protocol services such as Warp message relayers.
+
+## Specification
+
+Subnet EVM and Coreth currently handle Warp message signature requests from peers by decoding the `AppRequest` bytes (forwarded from Avalanchego via the `AppHandler` interface) using a type registered with the message codec:
+
+```
+type MessageSignatureRequest struct {
+	MessageID ids.ID `serialize:"true"`
+}
+```
+
+`MessageID` represents the VM-agnostic Warp message ID, and could be ported as-is to the protocol. To do so, we would [define](https://github.com/ava-labs/avalanchego/blob/v1.11.10-status-removal/proto/p2p/p2p.proto) a new P2P `Message` type:
+
+```diff
+message Message {
+    ...
+    // App messages:
+    AppRequest app_request = 30;
+    AppResponse app_response = 31;
+    AppGossip app_gossip = 32;
+    AppError app_error = 34;
+    
++   // Warp signature messages:
++   MessageSignatureRequest = 35;
+}
+```
+
+Where `MessageSignatureRequest` is defined identically to Subnet EVM/Coreth:
+```
+message MessageSignature {
+    bytes message_id = 1;
```

Comment:
I think we are going to also need a way to protect against unrequested responses hitting us. The sender can use something like a requester-maintained `request_id` field so that we drop messages for a `request_id` we don't know about, similar to `AppRequest` messages.

We will also need a way to route this message between VMs, as the p2p layer handles networking for the entire node we will need to have some field like `chain_id` so that we can route messages to the correct VM.

Reply:
Correct me if I'm wrong, but routing is no longer a concern since that is handled at the `AppRequest` level.